### DISCUSSION TITLE
Clerk as an OAuth Provider.

### DISCUSSION THREAD

**Discussion by mchennupati** [nextauthjs/next-auth#9316] [Ideas]
### Goals

1. Currently NextAuth allows Auth0 as an OAuth2 provider, can we also add an plug and play for Clerk as well
2. One could then avoid having to specify the custom template.
3. Allows for easy switching between competing providers.

### Non-Goals

1.  Simple addition of provider, no additional features required.

### Background

The current alternatives is to create your own configuration file in this form: 
https://github.com/nextauthjs/next-auth/tree/v4/packages/next-auth/src/providers

The clerk oauth2 can be created using this: 
https://api.clerk.com/v1/oauth_applications

And it is in this form:
{
    "object":"oauth_application",
    "id":"xxxxxx", 
    "instance_id":"ins_xxxx",
    "name":"nextauth-clerk",
     "client_id":"xxxxx",
    "client_secret":"xxxxxx",
    "public":false,
    "scopes":"email profile",
    "callback_url":"https://yourapp/api/auth",
    "authorize_url":"https://clerk.yourapp.com/oauth/authorize",
    "token_fetch_url":"https://clerk.yourapp.com/oauth/token",
    "user_info_url":"https://clerk.yourapp.com/oauth/userinfo",
    "created_at":1701767138671,
    "updated_at":1701767138671
}

### Proposal

Add a new clerk.ts file to providers here: https://github.com/nextauthjs/next-auth/tree/v4/packages/next-auth/src/providers

**Comment by MatteoGauthier** [comment]
Hey @mchennupati is the code you provided working?

**Reply by mchennupati** [reply]
Hey apologies for the delay, the short answer is I couldn't get around making it a custom package so I could import it. Cloning it, adding this line and getting a new npm package (the last bit was a but too much as I got entangled in type errors etc :))
So I ended up removing nextauth and just adding clerk.

**Reply by MatteoGauthier** [reply]
Thanks for your reply ! I choose to use [Logto](https://logto.io) as a provider, an open source authentification service with a lot of enterprise features, it's a Auth0 alternative.
Cheers

**Reply by masoncfrancis** [reply]
@MatteoGauthier have you been able to use Logto with NextAuth? I’m exploring authentication options and I really like it, but I’m trying to figure out what it would take to implement

**Comment by rogervila** [comment]
I have started a PR to implement Clerk as an OAuth provider (https://github.com/nextauthjs/next-auth/pull/11349).
Do not hesitate to review it.

**Reply by rogervila** [reply]
CC @mchennupati @masoncfrancis @MatteoGauthier

**Reply by MatteoGauthier** [reply]
Great news @rogervila thanks !

**Comment by mchennupati** [comment]
Thanks ! I shall take a look when I have a chance, I had switched to using
clerk on its own due to time constraints back then, but this will be much
appreciated by a number of people.

On Tue, Jul 9, 2024 at 12:24 AM Mattèo Gauthier ***@***.***>
wrote:

> Great news @rogervila <https://github.com/rogervila> thanks !
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextauthjs/next-auth/discussions/9316#discussioncomment-9992128>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA7RXOK73KMATJ7LKZKZF73ZLMGSNAVCNFSM6AAAAABAHLPH56VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TSOJSGEZDQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Mohanakrishna Chennupati
Mobile:      07774328123
